pkg/operator/controller/ingress/controller.go

```diff
@@ -566,6 +567,11 @@ func setDefaultPublishingStrategy(ic *operatorv1.IngressController, platformStat
 						changed = true
 					}
 				}
+				if statusLB.ProviderParameters.AWS.Type == operatorv1.AWSNetworkLoadBalancer {
+					if statusLB.ProviderParameters.AWS.NetworkLoadBalancerParameters == nil {
+						statusLB.ProviderParameters.AWS.NetworkLoadBalancerParameters = &operatorv1.AWSNetworkLoadBalancerParameters{}
+					}
```

Comment:
I think I agree with the status being updated for the state of the managed objects instead of the spec. Also, I agree that it may easily become messy and the source of truth may be confused.

This made me think of `updateIngressControllerAWSSubnetStatus` function from `status.go`. It relies on the LB type read from the status which in turn was set from the spec (`setDefaultPublishingStrategy`) .  May this cause a problem taking into account that we clean the other type's params (CLB's if NLB is set and vice versa)? I suppose that we can turn out into a situation when the type changed in the status but the service object failed to be updated - in that case we can prematurely clean up the LB params. Should we read the service type from the annotation too?

Reply:
> May this cause a problem taking into account that we clean the other type's params (CLB's if NLB is set and vice versa)? 

Okay, so this comment led me to attempt to enumerate what would happen, which made me realize there was a bug when I tried it out. We did have E2E for this, but E2E was passing "falsely" for this scenario. I have now fixed this issue and documented here: https://github.com/openshift/cluster-ingress-operator/pull/1046#issuecomment-2244030050 and https://github.com/openshift/cluster-ingress-operator/pull/1046#pullrequestreview-2194592829. This might be more than what you asked for.

This is *now* what happens when you change LB Type and NLB/CLB have different subnet values:
1. The IngressController triggers a reconcile loop, [readmits](https://github.com/openshift/cluster-ingress-operator/blob/a180755f2f4d082b3236ef88278cf360c52ea32d/pkg/operator/controller/ingress/controller.go#L288) the IngressController which updates the status to only what is set in `setDefaultPublishingStrategy` which "cleans" the status like you said.
2. It requeues, and runs reconcile again, which now runs `ensureIngressController`.
3. Now the `ensureLoadBalancerService` will change the LB Type, but also change the subnets as well per my [most recent update](https://github.com/openshift/cluster-ingress-operator/pull/1046/files#diff-5ee46ab285ecdbf9373a451d224299f443d5821af04eb5d0a7cee19ec0a78ed8R668-R672). Before, it didn't change the subnets, which caused many problems.
4. Then `syncIngressControllerStatus` runs, and updates the status with the correct subnets.

Hope that answers your question. Just in writing this out, I found another minor issue: the call to `computeLoadBalancerProgressingStatus` doesn't use the `updated` status, so it actually adds a `LoadBalancerProgressing` status momentarily because the subnet status isn't updated until `syncIngressControllerStatus` is finished:
https://github.com/openshift/cluster-ingress-operator/blob/a180755f2f4d082b3236ef88278cf360c52ea32d/pkg/operator/controller/ingress/status.go#L86

That might be okay, as it's temporary, but I need to look into if it's easily fixable.

> I suppose that we can turn out into a situation when the type changed in the status but the service object failed to be updated - in that case we can prematurely clean up the LB params. Should we read the service type from the annotation too?

Hmm. It's a valid point. `status.endpointPublishingStrategy.loadBalancer.providerParameters.aws.type` will be [updated regardless](https://github.com/openshift/cluster-ingress-operator/blob/3dc5456f0492e5a9bd7ec1de974813e58059101c/pkg/operator/controller/ingress/controller.go#L546) if the service failed to update. However, other logic will such as [IsProxyProtocolNeeded](https://github.com/openshift/cluster-ingress-operator/blob/a180755f2f4d082b3236ef88278cf360c52ea32d/pkg/operator/controller/ingress/controller.go#L1183) will break too if that happens.

I personally think the LB Type _should_ use the annotation for the real value, however, given it's an outstanding issue, I think we should solve it in another PR/bug/solution.